### ISSUE TITLE
[Diem Framework Feature Request] Concurrent Preburns

### ISSUE THREAD

**Issue by tzakian** [diem/diem#7552]
We want to add the ability to have concurrent preburns in the Diem Framework. 

The implementation should satisfy the following requirements:

* Current `Preburn` resources should still be able to be used for `TransactionFee` collection and burning.
* Multiple preburn requests in the same currency must be able to be outstanding at the same from the same designated dealer.
* Burning takes an amount `x` to burn along with the currency to be burned. The `Preburn` resource in the outstanding preburns for that currency are indexed by their value. The outstanding preburn resource with a value equal `x` and smallest index in the vector of outstanding preburn requests will be removed (i.e., preburns of the same value are processed in FIFO order). After this, the removed preburn resource will have its coins destroyed and the preburn resource will be destroyed.
* The same process is performed for burn cancellations as with burns, except that the coins are returned back instead of being burned.
* It is an error if a burn/cancel burn request does not find an outstanding preburn with a matching value.
* The maximum number of outstanding preburn requests in a currency for any designated dealer is 256. Exceeding this number is an error.
* Both a designated dealer and the treasury compliance account can publish a concurrent preburn resource under it.
* An account cannot have both a `ConcurrentPreburn` resource, and a `Preburn` resource for that same currency published under their account at the same time (i.e., these two resources are mutually exclusive).
* The ordering of requests in the `ConcurrentPreburn`s resource should remain stable.
    -  A new preburn request should have the greatest index in the `ConcurrentPreburn`s vector. 
    - A burn/cancellation should remove the matching preburn resource with the smallest index. The ordering of all other elements in the vector should remain unchanged. 
* The sum of all coins in a currency that exist in `Preburn` resource should be equal to the `preburn_value` field in the `CurrencyInfo` resource for that currency
* Preburning a coin in a currency should not change the `total_value` field in the `CurrencyInfo` resource for that currency.
* Burning a coin with a value of `x` in a currency should decrease the value of the `total_value` field in the `CurrencyInfo` resource for that currency by `x`. The `preburn_value` in that same `CurrencyInfo` resource should decrease by `x` as well.
* Cancelling a burn with a value of `x` in a currency should decrease the `preburn_value` field in that currency's `CurrencyInfo` resource by `x`. The `total_value` field in that same `CurrencyInfo` resource should remain unchanged.

**Comment by moeziniaf** [comment]
perfection! Also, several of the api's remain unchanged with this update such as: `preburn_with_resource`, `preburn_to`, `burn_with_capability`, `cancel_burn_with_capability` etc .. this will alleviate the pain of updating most of the specs in the module.

**Comment by tzakian** [comment]
Good point! @moeziniaf I've added another requirement at the end (in italics) that a preburn request needs to have a nonzero value. I don't think it should be an issue, but let me know if this would cause any problems.
